File: HackerRank/30 Days of Code/Day13_AbstractClasses.java
// Day 13: Abstract Classes
// https://www.hackerrank.com/challenges/30-abstract-classes/problem

import java.util.*;

abstract class Book {
    String title;
    String author;
    
    Book(String title, String author) {
        this.title = title;
        this.author = author;
    }
    
    abstract void display();
}

class MyBook extends Book{
    int price;
    
    MyBook(String t, String a, int p){
        super(t, a);
        this.price = p;
    }
    
    @Override
    public void display(){
        System.out.println("Title: " + title);
        System.out.println("Author: " + author);
        System.out.println("Price: " + price);
    }
}

public class Solution {
   
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String title = scanner.nextLine();
        String author = scanner.nextLine();
        int price = scanner.nextInt();
        scanner.close();

        Book book = new MyBook(title, author, price);
        book.display();
    }
}

// 추상 클래스를 상속받은 클래스 생성 및 사용
// Java에서 추상 메서드 생성 및 해당 클래스를 상속받아 실체화(@Override)

File: HackerRank/30 Days of Code/Day13_AbstractClasses.cpp
// Day 13: Abstract Classes
// https://www.hackerrank.com/challenges/30-abstract-classes/problem

#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <string>
using namespace std;
class Book{
    protected:
        string title;
        string author;
    public:
        Book(string t,string a){
            title=t;
            author=a;
        }
        virtual void display()=0;

};

class MyBook : Book{
        int price;
    public:
        MyBook(string t, string a, int p) : Book(t, a){
            price = p;
        }
        virtual void display(){
            cout << "Title: " << title << "\nAuthor: " << author << "\nPrice: " << price << '\n';
        }
};

int main() {
    string title,author;
    int price;
    getline(cin,title);
    getline(cin,author);
    cin>>price;
    MyBook novel(title,author,price);
    novel.display();
    return 0;
}

// 추상 클래스를 상속받은 클래스 생성 및 사용
// C++에서 추상 함수가 있는 클래스 상속 및 실체화